### PR TITLE
[ao][sparsity] Base scheduler class for Data Schedulers

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79883
* #79822
* __->__ #79817

The BaseDataScheduler is the abstract scheduler class specifically for the
BaseDataSparsifier class. This class controls a specific hyperparameter of
the sparsifier class and varies it across the training process (or across time).

Args:
    data_sparsifier (instance of BaseDataSparsifier)
        Implemented class data sparsifier class wherein the update_mask is implemented
    schedule_param (str)
        A specific hyperparameter of the passed sparsifier that needs to be scheduled/varied
    last_epoch (int, default=-1)
        This is specifically is passed when training needs to be resumed from a particular
        point.
    verbose (bool, default=False)
        Verbosity of the BaseDataScheduler

The *get_schedule_param()* function needs to be implemented by the user.

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataScheduler```

Differential Revision: [D37358608](https://our.internmc.facebook.com/intern/diff/D37358608)

### CODE REVIEW

**Comment by vadimkantorov** [comment]
It may be better to fix design issues in _LRScheduler than copying them in a new scheduler hierarchy: https://github.com/pytorch/pytorch/issues/68332

File: test/ao/sparsity/test_data_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    def _get_sparsifier(self):
+        data_list, defaults, data_with_config = self._get_data()
+        sparsifier = DataNormSparsifier(data_list, **defaults)
+        for data_config_dict in data_with_config:
+            name, data, config = data_config_dict['name'], data_config_dict['data'], data_config_dict['config']
+            sparsifier.add_data(name=name, data=data, **config)
+        return sparsifier
+
+    def _get_scheduler(self, sparsifier):
+        schedule_param = self._get_scheduler_param()
```
Comment (jerryzh168) [inline test/ao/sparsity/test_data_scheduler.py:50]: nit: it's not obvious that `_get_scheduler` depends on `_get_scheduler_param`, can we make scheduler param as an argument to this function to make it more explicit?

**Review by jerryzh168 — APPROVED** [review]
as we discussed offline, might be useful to think about how to merge with/reuse original scheduler/sparsifier in sparsity

**Review by z-a-f — APPROVED** [review]

File: test/ao/sparsity/test_data_scheduler.py (hunk 1 above)
Comment (macandro96) [reply test/ao/sparsity/test_data_scheduler.py]: makes sense - made changes

File: torch/ao/sparsity/experimental/data_scheduler/base_data_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        self.verbose = verbose
+
+        # Housekeeping
+        self._get_sp_called_within_step: bool = False
+
+    @abc.abstractmethod
+    def get_schedule_param(self):
+        r"""Abstract method that needs to be implemented by the child class.
+            Returns: Dictionary of mapping from name -> sparsifier_hyperparam value
+        """
```
[the remark this answers is not in the document]
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_scheduler/base_data_scheduler.py]: add an example docstring

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37358608).
